Reworked and fixed and security improved version

nugged opened the pull request:
Story:
- Prevent for undef error with language list in a case if not yet any language installed (koha_installed_languages is defined)
- Add place where to put domain in hosts (insertafter)
- Move server-related configs to be grouped
- White space: blocks visually grouped
- Add main repository only if key defined (koha_signing_key_id)
- Make main repository named "koha-community"
- Add feature to add key by content (koha_signing_key_content)
- Add feature to have secondary koha repository
- Feature for future re-use to install one custom built koha-common over general koha-common with dependencies. Vars:
koha_extra_keyserver, koha_extra_signing_key_id, koha_extra_signing_key_content, koha_extra_package_repository 
- Add pinning for secondary repo and koha_extra_package_rep_domain variable
- Add sshd hardening
- Add Apache security hardening

nugged committed:
Add place where to put domain in hosts
(insertafter)

nugged committed:
Move server-related configs to be grouped

nugged commented:
@kohaputti: split, described, updated & simplified, 

testing ongoing (still: had connectivity issues to the main deb repo for my VM from home, will try at the morning once again)

nugged committed:
Prevent for undef error with language list
in a case if not yet any language installed
(koha_installed_languages is defined)

nugged committed:
White space: blocks visually grouped

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
-    filename: koha-packages
+    filename: koha-community
+
+- name: Add Koha package our repository signing key
```
Here also, our -> secondary

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+    id: "{{ koha_extra_signing_key_id }}"
+  when: koha_extra_signing_key_id is defined and koha_extra_keyserver is defined
+
+- name: Add Koha package our repository signing key by content
```
Here also, our -> secondary. Maybe also just "Add secondary package repository signing key"

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+    data: "{{ koha_extra_signing_key_content }}"
+  when: koha_extra_signing_key_content is defined
+
+- name: Add Koha package our repository
```
Here also, our -> secondary.

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+- name: Add Koha package our repository signing key by content
+  apt_key:
+    data: "{{ koha_extra_signing_key_content }}"
+  when: koha_extra_signing_key_content is defined
```
I think the more common way to do this is just 
`when: koha_extra_signing_key_content`
i.e. without the is defined part. More background on this on the later review comment.

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+
+- name: Add Koha package our repository
+  apt_repository:
+    repo: "{{ koha_extra_package_repository }}"
```
extra -> secondary. Or if you prefer maybe all instances of secondary I mentioned above should also be extra? Either word is fine in my opinion, but just lets keep it consistent.

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+  apt_repository:
+    repo: "{{ koha_extra_package_repository }}"
+    state: present
+    filename: koha-kcminni
```
kcminni -> secondary

kohaputti commented on tasks/main.yml:
```diff
@@ -2,16 +2,58 @@
+    owner: root
+    group: root
+    mode: u=rw,g=r
+  when: koha_extra_package_rep_domain is defined
```
The name doesn't tell me clearly what this is about, I think this should be documented along with the other variables in README.md. Also here there is the "is defined" I mentioned earlier, e.g. https://github.com/geerlingguy/ansible-role-gitlab there is defaults/main.yml variables with just "" as the value so maybe we could do the same, but I also see it doing default(false) but that is just extra work in my opinion and the role user is unlikely to somehow undefine the variable. https://docs.ansible.com/ansible/latest/user_guide/playbooks_best_practices.html#keep-it-simple https://medium.com/opsops/is-defined-in-ansible-d490945611ae

nugged replied:
In a case variable will be not defined in file – script will fail if just to left "only" variable.
So I propose and did this in both-checks way, now it works properly in all cases:

```koha_signing_key_content is defined and koha_signing_key_content|length > 0```

(for example https://www.shellhacks.com/ansible-when-variable-is-defined-exists-empty-true/ https://stackoverflow.com/questions/36912726/how-to-test-that-a-registered-variable-is-not-empty/53362496 )

kohaputti commented on tasks/main.yml:
```diff
@@ -113,6 +172,30 @@
 
+
+
+- name: make in sshd_config PermitRootLogin to be off
```
Same as with apache hardening commit.

nugged replied:
removed but stored, I will push this hardening code in another role on which we will decide.

kohaputti commented on tasks/main.yml:
```diff
@@ -79,6 +120,22 @@
     mode: u=rw,g=r
   register: apache_conf
 
+- name: "Apache security: ensure 'ServerTokens' to be 'Prod'"
```
I'm inclined to drop this commit introducing apache hardening, we should do this in other role since this should be only about making koha work. This would be something that not only applies to Koha and might be useful for other applications running on the server.

nugged replied:
removed but stored, I will push this hardening code in another role on which we will decide.

nugged committed:
Make primary repository apt file named "koha-primary"

nugged committed:
Add feature to have secondary koha repository

this is for next commits (future implementation): to install
one custom koha-secondary deb over general koha-primary
with primary dependencies

Vars:
- koha_secondary_keyserver
- koha_secondary_signing_key_id
- koha_secondary_signing_key_content
- koha_secondary_package_repository

nugged committed:
Add pinning for secondary repo
and koha_secondary_package_rep_domain variable

nugged committed:
Add feature to add key by content (koha_signing_key_content)

nugged committed:
Add primary repository only if key defined
(koha_signing_key_id)

kohaputti commented on tasks/main.yml:
```diff
@@ -48,6 +48,15 @@
+    owner: root
+    group: root
+    mode: u=rw,g=r
+  when: koha_secondary_package_rep_domain is defined and koha_secondary_package_rep_domain|length > 0
```
is the length > 0 needed? Would

`when: koha_secondary_package_rep_domain is defined and koha_secondary_package_rep_domain`

be equivalent? Please add documentation also for this variable to the README file.

kohaputti replied:
length > 0 and just variable name has a bit different semantics so let's go with length > 0 given it works for "0" string too.
